### PR TITLE
feat: add a Gno2Go precompiler

### LINKED ISSUE #107: Doesn't check for return
Preprocessor doesn't check whether all code branch paths return in functions that have a return value.
This causes unexpected behavior during runtime, in my case I got:
`interface conversion: gno.Stmt is nil, not *gno.AssignStmt`. 

TODO: make preprocessor ensure all branch paths return when necessary.
Alternatively, use the go compiler for type-checking.

### PR DESCRIPTION
**Pull request by moul**
This PR is a proposal for https://github.com/gnolang/bounties#10-use-go-typechecker

Official target:

- [x] Rename .go to .gno where appropriate
- [x] Make gno precompiler (which comes before preprocessor) compile .gno to .go in same folder.
- [x] "gno.land/r/realm" in .gno becomes "github.com/gnolang/gno/examples/gno.land/r/realm" in .go.
- [x] "std" becomes "github.com/gnolang/gno/stdlibs/stdshim" (?)
- [x] keeps a whitelist of stdlib packages (to prevent attacks, to prevent compiling malware).
- [ ] Precompiler tries to compile the translated .go code using go, before proceeding.

---

My plan/implementation:

* [x] implement a transformation system in the `gno.Gno2Go` func.
* [x] replace `std` with `github.com/gnolang/gno/stdlibs/stdshim`.
* [x] replace `gno.land/r/realm` with `github.com/gnolang/gno/examples/gno.land/r/realm`.
* [x] always add a top-level comment to indicate that the .go files was generated (in case someone commits the generated files).
* [x] removed `tests/go.mod`, no you can safely run `go test -v ./...` from the top-level directory.
* [x] because `./stdlibs/*` is both used by the `.gno` files and also by some `.go` files, I decided to rename them to `.gno` and create symlinks for each files so that `go test ./...` can work.
* [x] initial version of a `gnodev` tool with `gnodev precompile ./path/to/package`
* [ ] bootstrap a simple `stdshim` package.
* [ ] way to ensure the go version we're checking against, for distributed determinism. (https://github.com/gnolang/gno/pull/119#issuecomment-1101902599)

---

~Also, I’ve a suggestion: what do about creating a new `cmd/gnodev` command, dedicated to contract developers, with tools like `precompile`, `test` and other developer tools?
That’s personally what I would love to work with.~

Note: the PR is pretty big, because it renames a lot of .go files to .gno; the interesting files here are `gno2go{,_test}.go`.

---

Fixes #107
Related with #121
Closes #113 (replace)

### CODE REVIEW

**Comment by jaekwon** [comment]
thank you @moul, this looks awesome, though i've yet to read the code yet. I will review by tomorrow when I'm back.
cmd/gnodev command sounds great.
I noticed that in the last commit, git didn't seem to register that many of the files were renamed. Can that be fixed, to preserve file history of the original .go files?

**Comment by moul** [comment]
> I noticed that in the last commit, git didn't seem to register that many of the files were renamed. Can that be fixed, to preserve file history of the original .go files?

Sure thing, I'll create a commit for the `git mv` and another one for the porting changes.

**Comment by moul** [comment]
I'm not sure why, but even with a dedicated commit for moving, and another one for fixing, GitHub seems to make shortcuts and only look the general diff with `master`.

---

here it works (probably because the change is small)

Screenshot 2022-04-17 at 20 35 05@2x

here it doesn't work

Screenshot 2022-04-17 at 20 35 23@2x

---

Any suggestions?

**Comment by moul** [comment]
New: an initial version of `gnodev`.

Screenshot 2022-04-18 at 01 12 00@2x

File: gno2go.go [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+var stdlibWhitelist = []string{
+	"regexp",
+	"std",
+	"strconv",
+	"strings",
+	"maths",
+}
+
+func Gno2Go(fset *token.FileSet, f *ast.File) (string, error) {
```
Comment (jaekwon) [inline gno2go.go:32]: To disambiguate from Go2Gno which works on values (and there may be a need for Gno2Go for values too), maybe this should be renamed to Gno2GoPrecompile.

Comment (moul) [reply gno2go.go]: fixed

File: tests/files/redeclaration-global0.gno [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package main
```
Comment (jaekwon) [inline tests/files/redeclaration-global0.gno:1]: spacing issue?

**Comment by jaekwon** [comment]
looks good so far... we probably also want a way to ensure the go version we're checking against, for distributed determinism.

File: tests/files/redeclaration-global0.gno (hunk 2 above)
Comment (moul) [reply tests/files/redeclaration-global0.gno:1]: Fixed, with `gofmt -w` (Just kept files/redeclaration2.gno unmodified on purpose).

**Comment by moul** [comment]
> looks good so far... we probably also want a way to ensure the go version we're checking against, for distributed determinism.

I added it in the list of checkboxes at the top of the PR, so we can remind to do them in a new one.

**Review by jaekwon — APPROVED** [review]

**Comment by jaekwon** [comment]
Merging in to prevent conflicts; another PR will complete the precompile logic. Thank you @moul.

**Pull request merged**